kafka/producer.go

```diff
@@ -212,6 +219,10 @@ func (p *Producer) ProcessBatch(ctx context.Context, batch *model.Batch) error {
 			return fmt.Errorf("failed to encode event: %w", err)
 		}
 		record.Value = encoded
+		if !p.cfg.Sync {
+			// Detach the context from its deadline or cancellation.
+			ctx = queuecontext.DetachedContext(ctx)
```

Comment:
Probably a stupid question, but why do we need this? I would think that if the client closes the connection we should abort the publish.

Reply:
Never mind, I think I understand now.

Reply:
It's got to do with the `client.Publish(ctx)` actually happening after the context may have been cancelled. We had a similar problem in the APM Server Intake calls with async set.